File: Voci/Sources/Model/TaskItem.swift
// Sources/Model/TaskItem.swift — UI/domain task model + mapping into VociCore.Task (v2 shape,
// contracts/vocicore-api.md). See VociTask.swift for the paired SwiftData persistence model.
import Foundation
import VociCore

/// Alias so app code (and Phase-2 views) can write `TaskStatus` without importing `VociCore`
/// everywhere — the lifecycle enum itself lives in the engine package and is not redefined here.
typealias TaskStatus = VociCore.TaskStatus

/// UI-facing priority, 1 (highest) ... 3 (lowest). Maps 1:1 onto `VociCore.Task.priority`'s
/// `Int` scale via `rawValue` (the engine additionally allows 4 / `nil`; this app never produces
/// those, so the UI enum only needs high/medium/low).
enum Priority: Int, Sendable, Equatable, CaseIterable, Codable {
    case high = 1
    case medium = 2
    case low = 3
}

/// UI-only "Now" vs "Later today" bucket. The engine has no notion of this; it is purely how
/// `TodayView` (Phase 2) groups the list, mirroring the prototype's `task.when`.
enum When: Sendable, Equatable, Codable {
    case now
    case later
}

struct TaskItem: Identifiable, Sendable, Equatable {
    let id: UUID
    var title: String
    var details: String
    var priority: Priority
    var status: TaskStatus
    var deadline: Date?
    /// Gates eligibility (replaces v1's `dependsOn: [UUID]` — data-model.md "Persisted layer").
    /// AND semantics, mirrors `VociCore.Task.conditions` 1:1 (see `snapshot()`). `.taskDone`
    /// edges must be validated via `TaskStore` before being attached (validation rule 1); this
    /// struct itself carries no validation, matching `VociCore.Condition`'s own pure-value
    /// nature — see `TaskStore`'s doc comments for the trust-boundary note.
    var conditions: [VociCore.Condition]
    var createdAt: Date
    var when: When
    var durationMinutes: Int?
    var frog: Bool

    // MARK: - v2 fields (specs/002-workflow-command-center/data-model.md "Persisted layer")

    /// Free-text notes, distinct from `details` (which stays the voice-capture read-back copy
    /// spoken by `AppState.speakDetails`).
    var notes: String?
    /// Verbatim utterance for voice-created tasks — always kept (FR-001).
    var sourceTranscript: String?
    var kind: TaskKind
    /// Leaf tasks only — `TaskStore.setRecurrence`/`add`/`addBatch` reject setting this on a task
    /// that has children (validation rule 2).
    var recurrence: Recurrence?
    /// `nil` -> global default `ReminderPolicy.defaultPolicy`.
    var reminderOverride: ReminderPolicy?
    /// "Save game" note surfaced on re-entry (FR-042).
    var resumeNote: String?
    /// Đổi-gió counter; ≥3 triggers a one-time breakdown suggestion (FR-030).
    var switchAwayCount: Int
    /// Latest completion instant; full history lives in `CompletionEvent` — recurrence resets
    /// in place, so this field alone can't reconstruct history.
    var completedAt: Date?
    /// Breakdown child -> parent link. Engine rule: a task that is a parent of any open child is
    /// excluded from `nextTask`, even though it stays selectable in its own right once every
    /// child is done/archived.
    var parentId: UUID?
    /// Present while delegated-to-AI (US4); cleared once the matching `.external` condition is
    /// satisfied (signal or manual).
    var delegation: DelegationMeta?

    init(
        id: UUID = UUID(),
        title: String,
        details: String = "",
        priority: Priority,
        status: TaskStatus = .todo,
        deadline: Date? = nil,
        conditions: [VociCore.Condition] = [],
        createdAt: Date = Date(),
        when: When,
        durationMinutes: Int? = nil,
        frog: Bool = false,
        notes: String? = nil,
        sourceTranscript: String? = nil,
        kind: TaskKind = .task,
        recurrence: Recurrence? = nil,
        reminderOverride: ReminderPolicy? = nil,
        resumeNote: String? = nil,
        switchAwayCount: Int = 0,
        completedAt: Date? = nil,
        parentId: UUID? = nil,
        delegation: DelegationMeta? = nil
    ) {
        self.id = id
        self.title = title
        self.details = details
        self.priority = priority
        self.status = status
        self.deadline = deadline
        self.conditions = conditions
        self.createdAt = createdAt
        self.when = when
        self.durationMinutes = durationMinutes
        self.frog = frog
        self.notes = notes
        self.sourceTranscript = sourceTranscript
        self.kind = kind
        self.recurrence = recurrence
        self.reminderOverride = reminderOverride
        self.resumeNote = resumeNote
        self.switchAwayCount = switchAwayCount
        self.completedAt = completedAt
        self.parentId = parentId
        self.delegation = delegation
    }

    /// Derived: "done" is purely the engine status, never a separately stored bool.
    var done: Bool { status == .done }

    /// Derived: formatted deadline time ("11:30 AM"); `nil` once done or when there's no
    /// deadline at all — matches the prototype's `timeBadge` (only ever shown for open tasks).
    var timeBadge: String? {
        guard !done, let deadline else { return nil }
        return deadline.formatted(.dateTime.hour().minute())
    }

    /// Derived: "45 min" / "1 hr" / "1h 30m" — matches the prototype's `dur` label.
    var durationLabel: String? {
        guard let durationMinutes, durationMinutes > 0 else { return nil }
        if durationMinutes < 60 { return "\(durationMinutes) min" }
        let hours = durationMinutes / 60
        let minutes = durationMinutes % 60
        if minutes == 0 { return hours == 1 ? "1 hr" : "\(hours) hrs" }
        return "\(hours)h \(minutes)m"
    }

    /// The only bridge into the pure selection engine (contracts/vocicore-api.md): maps this UI
    /// struct into the small value type `nextTask(from:now:calendar:)` actually consumes. No selection
    /// logic is duplicated here. `durationMinutes` feeds the engine's `estimateMinutes` — same
    /// concept, kept under its existing UI-facing name here (see TaskStore migration notes for
    /// why the persisted attribute name wasn't renamed to match).
    func snapshot() -> VociCore.Task {
        VociCore.Task(
            id: id,
            title: title,
            status: status,
            priority: priority.rawValue,
            deadline: deadline,
            conditions: conditions,
            estimateMinutes: durationMinutes,
            parentId: parentId,
            createdAt: createdAt
        )
    }
}

File: windows/src/Volar.Domain/TaskItem.cs
// Volar.Domain/TaskItem.cs — port of Sources/Model/TaskItem.swift: UI/domain task model + mapping
// into Volar.Core.TaskSnapshot (v2 shape, contracts/volarcore-api.md).
using System.Globalization;
using Volar.Core;

namespace Volar.Domain;

/// <summary>
/// UI-facing priority, 1 (highest) ... 3 (lowest). Maps 1:1 onto <see cref="TaskSnapshot.Priority"/>'s
/// <c>int?</c> scale via the numeric enum value (the engine additionally allows 4 / <see
/// langword="null"/>; this app never produces those, so this enum only needs high/medium/low).
/// </summary>
public enum Priority
{
    High = 1,
    Medium = 2,
    Low = 3
}

/// <summary>
/// UI-only "Now" vs "Later today" bucket. Volar.Core has no notion of this; it is purely how the
/// app groups its task list, mirroring the prototype's <c>task.when</c>.
/// </summary>
public enum When
{
    Now,
    Later
}

/// <summary>
/// UI/domain task model — the mutable value the app edits and lists, distinct from
/// <see cref="TaskSnapshot"/> (the small immutable-by-convention value <see cref="Snapshot"/> maps
/// into for the selection engine).
/// </summary>
/// <remarks>
/// Declared as a (non-<see langword="readonly"/>) <see langword="record struct"/> — like
/// <see cref="TaskSnapshot"/> — to mirror the Swift <c>struct TaskItem: Identifiable, Sendable,
/// Equatable</c>'s <c>var</c>-field (mutable-in-place, copy-on-assignment) value semantics.
/// <para>
/// Unlike the Swift source, <see cref="Id"/> and <see cref="CreatedAt"/> have no default values in
/// the constructor below: Swift defaulted them to <c>UUID()</c> / <c>Date()</c>, but the latter
/// reads the system clock, which this project's domain-purity rule forbids ("mọi thời điểm 'bây
/// giờ' phải là tham số truyền vào") — every instant must be supplied explicitly by the caller.
/// </para>
/// </remarks>
public record struct TaskItem
{
    public Guid Id { get; set; }
    public string Title { get; set; }
    public string Details { get; set; }
    public Priority Priority { get; set; }
    public TaskState Status { get; set; }
    public DateTimeOffset? Deadline { get; set; }

    /// <summary>
    /// Gates eligibility (replaces v1's <c>dependsOn: [Guid]</c> — data-model.md "Persisted
    /// layer"). AND semantics, mirrors <see cref="TaskSnapshot.Conditions"/> 1:1 (see
    /// <see cref="Snapshot"/>). A <see cref="TaskDoneCondition"/> must be validated via
    /// <see cref="DependencyGraph.ValidateCondition"/> by the persistence layer before being
    /// attached — this struct itself carries no validation, matching <see cref="Condition"/>'s own
    /// pure-value nature.
    /// </summary>
    public IReadOnlyList<Condition> Conditions { get; set; }

    public DateTimeOffset CreatedAt { get; set; }
    public When When { get; set; }
    public int? DurationMinutes { get; set; }
    public bool Frog { get; set; }

    // MARK: - v2 fields (specs/002-workflow-command-center/data-model.md "Persisted layer")

    /// <summary>Free-text notes, distinct from <see cref="Details"/> (the voice-capture read-back copy).</summary>
    public string? Notes { get; set; }

    /// <summary>Verbatim utterance for voice-created tasks — always kept (FR-001).</summary>
    public string? SourceTranscript { get; set; }

    public TaskKind Kind { get; set; }

    /// <summary>
    /// Leaf tasks only — the persistence layer rejects setting this on a task that has children
    /// (data-model.md validation rule 2).
    /// </summary>
    public Recurrence? Recurrence { get; set; }

    /// <summary><see langword="null"/> -&gt; global default <see cref="ReminderPolicy.DefaultPolicy"/>.</summary>
    public ReminderPolicy? ReminderOverride { get; set; }

    /// <summary>"Save game" note surfaced on re-entry (FR-042).</summary>
    public string? ResumeNote { get; set; }

    /// <summary>Switch-away counter; &gt;= 3 triggers a one-time breakdown suggestion (FR-030).</summary>
    public int SwitchAwayCount { get; set; }

    /// <summary>
    /// Latest completion instant; full history lives elsewhere (a <c>CompletionEvent</c> log,
    /// outside this project's scope) — recurrence resets in place, so this field alone can't
    /// reconstruct history.
    /// </summary>
    public DateTimeOffset? CompletedAt { get; set; }

    /// <summary>
    /// Breakdown child -&gt; parent link. Engine rule: a task that is a parent of any open child is
    /// excluded from <see cref="NextTaskSelector.NextTask"/>, even though it stays selectable in its
    /// own right once every child is done/archived.
    /// </summary>
    public Guid? ParentId { get; set; }

    /// <summary>
    /// Present while delegated-to-AI (US4); cleared once the matching <see cref="ExternalCondition"/>
    /// is satisfied (signal or manual).
    /// </summary>
    public DelegationMeta? Delegation { get; set; }

    public TaskItem(
        Guid id,
        string title,
        Priority priority,
        When when,
        DateTimeOffset createdAt,
        string details = "",
        TaskState status = TaskState.Todo,
        DateTimeOffset? deadline = null,
        IReadOnlyList<Condition>? conditions = null,
        int? durationMinutes = null,
        bool frog = false,
        string? notes = null,
        string? sourceTranscript = null,
        TaskKind kind = TaskKind.Task,
        Recurrence? recurrence = null,
        ReminderPolicy? reminderOverride = null,
        string? resumeNote = null,
        int switchAwayCount = 0,
        DateTimeOffset? completedAt = null,
        Guid? parentId = null,
        DelegationMeta? delegation = null)
    {
        Id = id;
        Title = title;
        Details = details;
        Priority = priority;
        Status = status;
        Deadline = deadline;
        Conditions = conditions ?? Array.Empty<Condition>();
        CreatedAt = createdAt;
        When = when;
        DurationMinutes = durationMinutes;
        Frog = frog;
        Notes = notes;
        SourceTranscript = sourceTranscript;
        Kind = kind;
        Recurrence = recurrence;
        ReminderOverride = reminderOverride;
        ResumeNote = resumeNote;
        SwitchAwayCount = switchAwayCount;
        CompletedAt = completedAt;
        ParentId = parentId;
        Delegation = delegation;
    }

    /// <summary>Derived: "done" is purely the engine status, never a separately stored bool.</summary>
    public readonly bool Done => Status == TaskState.Done;

    /// <summary>
    /// Derived: formatted deadline time ("11:30 AM"); <see langword="null"/> once done or when
    /// there's no deadline at all — matches the prototype's <c>timeBadge</c> (only ever shown for
    /// open tasks). Formats whatever offset <see cref="Deadline"/> itself carries (this type has no
    /// injected <see cref="TimeZoneInfo"/> of its own); always <see cref="CultureInfo.InvariantCulture"/>
    /// so the rendered string never varies by machine locale.
    /// </summary>
    public readonly string? TimeBadge
    {
        get
        {
            if (Done || Deadline is not DateTimeOffset deadline)
            {
                return null;
            }
            return deadline.ToString("h:mm tt", CultureInfo.InvariantCulture);
        }
    }

    /// <summary>Derived: "45 min" / "1 hr" / "1h 30m" — matches the prototype's <c>dur</c> label.</summary>
    public readonly string? DurationLabel
    {
        get
        {
            if (DurationMinutes is not int minutes || minutes <= 0)
            {
                return null;
            }
            if (minutes < 60)
            {
                return $"{minutes.ToString(CultureInfo.InvariantCulture)} min";
            }
            var hours = minutes / 60;
            var remainder = minutes % 60;
            if (remainder == 0)
            {
                return hours == 1 ? "1 hr" : $"{hours.ToString(CultureInfo.InvariantCulture)} hrs";
            }
            return $"{hours.ToString(CultureInfo.InvariantCulture)}h {remainder.ToString(CultureInfo.InvariantCulture)}m";
        }
    }

    /// <summary>
    /// The only bridge into the pure selection engine (contracts/volarcore-api.md): maps this
    /// domain struct into the small value type <see cref="NextTaskSelector.NextTask"/> actually
    /// consumes. No selection logic is duplicated here. <see cref="DurationMinutes"/> feeds the
    /// engine's <see cref="TaskSnapshot.EstimateMinutes"/> — same concept, kept under its existing
    /// domain-facing name here.
    /// </summary>
    public readonly TaskSnapshot Snapshot() => new(
        Id,
        Title,
        Status,
        (int)Priority,
        Deadline,
        Conditions,
        DurationMinutes,
        ParentId,
        CreatedAt);
}
